src/app/dataset-view/additional-components/lineage-component/services/lineage-graph-builder.service.ts

```diff
@@ -0,0 +1,163 @@
+import { Injectable } from "@angular/core";
+import { Observable, Subject, combineLatest, map, of, switchMap } from "rxjs";
+import { DatasetLineageBasicsFragment, DatasetKind, FetchStepUrl } from "src/app/api/kamu.graphql.interface";
+import { LineageUpdate } from "src/app/dataset-view/dataset.subscriptions.interface";
+import { DatasetSubscriptionsService } from "src/app/dataset-view/dataset.subscriptions.service";
+import { LineageGraphNodeKind, LineageGraphNodeData, LineageNodeAccess, LineageGraph } from "../lineage-model";
+import { Node, Edge } from "@swimlane/ngx-graph";
+
+@Injectable({
+    providedIn: "root",
+})
+export class LineageGraphBuilderService {
+    private lineageSourceLinkChanges: Subject<Edge[]> = new Subject<Edge[]>();
+
+    public changeSourceLink(links: Edge[]): void {
+        this.lineageSourceLinkChanges.next(links);
+    }
+
+    public get onLineageSourceLinkChanges(): Observable<Edge[]> {
+        return this.lineageSourceLinkChanges.asObservable();
+    }
+
+    constructor(private datasetSubsService: DatasetSubscriptionsService) {}
+
+    public getCurrentDataset(): Observable<DatasetLineageBasicsFragment> {
+        return this.datasetSubsService.onLineageDataChanges.pipe(map((data: LineageUpdate) => data.origin));
+    }
+
+    public buildGraph(): Observable<LineageGraph> {
+        return combineLatest([this.buildGraphNodes(), this.buildGraphLinks()]).pipe(
+            switchMap(([nodes, links]: [Node[], Edge[]]) => of({ nodes, links })),
+        );
+    }
+
+    private buildSourceNodes(): Observable<Node[]> {
+        return this.datasetSubsService.onLineageDataChanges.pipe(
+            map((data: LineageUpdate) => {
+                const uniqueDatasets: Record<string, DatasetLineageBasicsFragment> = {};
+                data.edges.forEach((edge: DatasetLineageBasicsFragment[]) =>
+                    edge.forEach((dataset: DatasetLineageBasicsFragment) => {
+                        uniqueDatasets[dataset.id] = dataset;
+                    }),
+                );
```

Comment:
You have the same two iterators to create `uniqueDatasets`, maybe it will be better to have a function for this?
It will be easier to re-use it or to change it

Reply:
Yes. I fixed this piece of code